contracts/zero-ex/contracts/src/features/UniswapFeature.sol

```diff
@@ -0,0 +1,320 @@
+                }
+
+                let receiver
+                // Is this the last pair contract?
+                switch eq(add(i, 1), numPairs)
+                    case 0 {
+                        // Not the last pair contract, so forward bought tokens to
+                        // the next pair contract.
+                        nextPair := computePairAddress(
+                            buyToken,
+                            loadTokenAddress(add(i, 2))
+                        )
+                        receiver := nextPair
+                    }
+                    default {
+                        // The last pair contract.
+                        // Forward directly to taker UNLESS they want ETH back.
+                        switch eq(buyToken, ETH_TOKEN_ADDRESS_32)
+                            case 0 {
+                                receiver := caller()
+                            }
+                            default {
+                                receiver := address()
+                            }
+                    }
+
+                // Call pair.swap()
+                mstore(0xB00, UNISWAP_PAIR_SWAP_CALL_SELECTOR_32)
+                switch pairOrder
+                    case 0 {
+                        mstore(0xB04, buyAmount)
+                        mstore(0xB24, 0)
+                    }
+                    default {
+                        mstore(0xB04, 0)
+                        mstore(0xB24, buyAmount)
+                    }
+                mstore(0xB44, receiver)
+                mstore(0xB64, 0x80)
+                mstore(0xB84, 0)
+                if iszero(call(gas(), pair, 0, 0xB00, 0xA4, 0, 0)) {
+                    bubbleRevert()
+                }
+            } // End for-loop.
+
+            // If buying ETH, unwrap the WETH first
+            if eq(buyToken, ETH_TOKEN_ADDRESS_32) {
+                // Call `WETH.withdraw(buyAmount)`
+                mstore(0xB00, WETH_WITHDRAW_CALL_SELECTOR_32)
+                mstore(0xB04, buyAmount)
+                if iszero(call(gas(), mload(0xA40), 0, 0xB00, 0x24, 0x00, 0x0)) {
+                    bubbleRevert()
+                }
+                // Transfer ETH to the caller.
+                if iszero(call(2300, caller(), buyAmount, 0xB00, 0x0, 0x00, 0x0)) {
+                    bubbleRevert()
+                }
+            }
+
+            // Functions ///////////////////////////////////////////////////////
+
+            function loadTokenAddress(idx) -> addr {
+                addr := and(ADDRESS_MASK, calldataload(add(mload(0xA00), mul(idx, 0x20))))
+            }
+
+            function normalizeToken(token) -> normalized {
+                normalized := token
+                // Translate ETH pseudo-tokens to WETH.
+                if eq(token, ETH_TOKEN_ADDRESS_32) {
+                    normalized := mload(0xA40)
+                }
+            }
+
+            function computePairAddress(tokenA, tokenB) -> pair {
```

Comment:
Could we get some better comments here? (Let me know if you want me to write some.)

In particular, writing to 0xB14 followed by 0xB00 and then hashing starting at 0xB0C won't be obvious to many people. I did some math and drew a diagram to make sure I understood what was where in memory. It would also be nice to explain the rest of the `CREATE2` address derivation, or perhaps just link to where Uniswap does/documents it.